lib/ClangImporter/ClangImporter.cpp

```diff
@@ -7717,26 +7722,44 @@ ValueDecl *
 CxxRecordAsSwiftType::evaluate(Evaluator &evaluator,
                                CxxRecordSemanticsDescriptor desc) const {
   auto cxxDecl = dyn_cast<clang::CXXRecordDecl>(desc.decl);
-  if (!cxxDecl)
+  // TODO: support specializations like swift::Optional<int>
+  if (!cxxDecl || isa<clang::ClassTemplateSpecializationDecl>(cxxDecl))
     return nullptr;
-  if (!isSwiftClassType(cxxDecl))
+  if (!importer::isSwiftType(cxxDecl))
     return nullptr;
 
-  SmallVector<ValueDecl *, 1> results;
+  if (!isSwiftClassType(cxxDecl)) {
+    // TODO: support value types indirected through swift::_impl::OpaqueStorage.
+    bool isOpaqueLayout = false;
+    for (const auto *field : cxxDecl->fields()) {
+      if (field->getName() == "_storage") {
+        isOpaqueLayout = !field->getType()->isArrayType();
+        break;
+      }
+    }
+    if (isOpaqueLayout)
+      return nullptr;
```

Comment:
I don't love this attempt to pattern-match exactly what we produce in export, especially because I know that export needs to be producing different patterns than this in some cases, like when a struct contains `float`s.  But if you're going to do it, you should really do it properly: check that there's exactly one field, check that the field has the name you expect, and check that its type is exactly a constant-size array of `char`.

Reply:
Indeed, I am also not the fan of pattern matching on the contents of the record. I looked into an alternative approach, but it required significantly more plumbing (need to heap allocate an `IRABIDetailsProvider` somewhere and plumb through a pointer a couple of calls). I figured since we plan to support round-tripping for all types, we might prefer the simpler check for a temporary solution. Let me know if that is not the case. In the meantime I made the pattern matching stricter as you suggested.